ec/ecresource/projectresource/elasticsearch.go

```diff
@@ -0,0 +1,364 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package projectresource
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/elastic/terraform-provider-ec/ec/internal/gen/serverless"
+	"github.com/elastic/terraform-provider-ec/ec/internal/gen/serverless/resource_elasticsearch_project"
+	"github.com/elastic/terraform-provider-ec/ec/internal/util"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+)
+
+func NewElasticsearchProjectResource() *Resource[resource_elasticsearch_project.ElasticsearchProjectModel] {
+	return &Resource[resource_elasticsearch_project.ElasticsearchProjectModel]{
+		modelHandler: elasticsearchModelReader{},
+		api: elasticsearchApi{
+			sleeper: realSleeper{},
+		},
+		name: "elasticsearch",
+	}
+}
+
+type elasticsearchModelReader struct{}
+
+func (es elasticsearchModelReader) Schema(ctx context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = resource_elasticsearch_project.ElasticsearchProjectResourceSchema(ctx)
+}
+
+func (es elasticsearchModelReader) ReadFrom(ctx context.Context, getter modelGetter) (*resource_elasticsearch_project.ElasticsearchProjectModel, diag.Diagnostics) {
```

Comment:
IMO we can replace it with a generic, e.g. 
```
func readFrom[T any](ctx context.Context, getter modelGetter) (*T, diag.Diagnostics) {
	var model *T
	diags := getter.Get(ctx, &model)
	return model, diags
}

func (es elasticsearchModelReader) ReadFrom(ctx context.Context, getter modelGetter) (*resource_elasticsearch_project.ElasticsearchProjectModel, diag.Diagnostics) {
  return readFrom[resource_elasticsearch_project.ElasticsearchProjectModel](ctx, getter)
}
```

Reply:
:) I was on the fence about whether this was worthwhile. You + part of me beats the other part of me though.